src/platform/mock-local-home-platform.ts

```diff
@@ -1,144 +1,91 @@
-  broadcastAddress: string;
-  listenPort: number;
-  broadcastPort: number;
-  discoveryPacket: string;
-  constructor(
-    state: ScanState,
-    broadcastAddress: string,
-    listenport: number,
-    broadcastPort: number,
-    discoveryPacket: string
-  ) {
-    this.state = state;
-    this.broadcastAddress = broadcastAddress;
-    this.broadcastPort = broadcastPort;
-    this.listenPort = listenport;
-    this.discoveryPacket = discoveryPacket;
-  }
-}
+export const ERROR_UNDEFINED_APP: string =
+  'Cannot trigger IdentifyRequest: App was undefined';
+export const ERROR_LISTEN_NOT_CALLED: string =
+  'Cannot trigger IdentifyRequest: listen() was not called';
+export const ERROR_UNDEFINED_IDENTIFYHANDLER: string =
+  'Cannot trigger IdentifyRequest: No Identify handler has been set by the fulfillment app';
+export const ERROR_UNDEFINED_VERIFICATIONID: string =
+  'The handler returned an IdentifyResponse with an undefined verificationId';
 
 // TODO(cjdaly): add other radio scan support
-export class MockLocalHomePlatform implements MockUDPListener {
+export class MockLocalHomePlatform {
   //  Singleton instance
   private static instance: MockLocalHomePlatform;
 
-  private udpScanConfigs: UDPScanConfig[] = [];
-  private mockNetwork: MockNetwork | undefined;
   private deviceManager: smarthome.DeviceManager = new DeviceManagerStub();
   private app: AppStub | undefined;
   private localDeviceIds: Map<string, string> = new Map<string, string>();
   private newDeviceRegisteredActions: ((localDeviceId: string) => void)[] = [];
-  private homeAppReady: boolean = false;
 
   private constructor() {}
 
-  public initializeRadio(
-    mockNetwork: MockNetwork,
-    udpScanConfigs: UDPScanConfig[]
-  ) {
-    this.mockNetwork = mockNetwork;
-    this.udpScanConfigs = udpScanConfigs;
-    this.setupUDP();
-  }
-
   public setApp(app: AppStub) {
     this.app = app;
   }
 
-  private onNewDeviceIdRegistered(localDeviceId: string) {
-    this.newDeviceRegisteredActions.forEach(newDeviceRegisteredAction => {
-      newDeviceRegisteredAction(localDeviceId);
-    });
-  }
-
-  public async getNextDeviceIdRegistered(): Promise<string> {
-    return new Promise(resolve => {
-      this.newDeviceRegisteredActions.push(localDeviceId => {
-        resolve(localDeviceId);
-      });
-    });
-  }
-
-  public isHomeAppReady(): boolean {
-    return this.homeAppReady;
+  public getDeviceManager(): smarthome.DeviceManager {
+    return this.deviceManager;
   }
 
-  public notifyHomeAppReady(): void {
-    this.homeAppReady = true;
+  public getLocalDeviceIdMap(): Map<string, string> {
+    return this.localDeviceIds;
   }
 
-  //  Singleton getter
-  public static getInstance(): MockLocalHomePlatform {
-    if (!MockLocalHomePlatform.instance) {
+  /**
+   * Returns the static singleton instance, creating it if needed.
+   * @param resetState  Whether or not to force a new instance.  Useful for initializing tests.
+   */
+  public static getInstance(
```

Comment:
Curious if this need to be a singleton?

Reply:
It's a singleton so an arbitrary `AppStub` instance can access the single instance of `MockLocalHomePlatform`.  This instance also needs to be accessed in tests in order to trigger intents and assert against.  Another option I was considering is to have a static event that fires when an `AppStub` is created.  Then a `MockLocalHomePlatform` could listen for that event and create the association.  That seems functionally the same as the current implementation, but would get rid of the "reset instance" flag.

The other option is to have the user re-implement their app (`App` constructor, setting handlers, call to `listen()`) within their test scope, so they can inject the dependency in that scope.

Open to suggestions.

Reply:
> It's a singleton so an arbitrary AppStub instance can access the single instance of MockLocalHomePlatform

Something to consider is if you'd ever need to associate multiple applications to a single platform stub instance.
If that's not the case maybe you could simply create a throwaway platform stub along side the app and keep a cross reference?

> This instance also needs to be accessed in tests in order to trigger intents and assert against. 

Could the tests import and explicit create a MockPlatform instance per test? That would also dispense you from implementing any kind of reset mechanism.

> The non-static option is to have the user re-implement their App creation (App constructor, setting handlers, call to listen()) within their test scope, so they can inject the dependency in that scope.

Another element is how explicit you want to make the platform stubbing from tests, one way to surface it would be that once imported/initialized `smarthome.App` is exposed in lieu of `AppStub` just like if you imported the real thing.

Reply:
> Something to consider is if you'd ever need to associate multiple applications to a single platform stub instance.
If that's not the case maybe you could simply create a throwaway platform stub along side the app and keep a cross reference?

There should only ever be one instance of the `App` and one instance of the `MockLocalHomePlatform`.  However, the intent is that `App` gets created in the developer's default code (for example, [this file](https://github.com/actions-on-google/smart-home-local/blob/master/app/index.ts)), while the `MockLocalHomePlatform` gets created in the testing scope.  The static instance is the cross-referencing mechanism.  If I created the `MockLocalHomePlatform` within the `App` stub, I couldn't access either from a test (without some other static mechanism)

> Could the tests import and explicit create a MockPlatform instance per test? That would also dispense you from implementing any kind of reset mechanism.

Yes, but I would need some way of notifying the next `App` stub of this instance.

> Another element is how explicit you want to make the platform stubbing from tests, one way to surface it would be that once imported/initialized smarthome.App is exposed in lieu of AppStub just like if you imported the real thing.

Good point, and something I'll change in this PR.

I think my takeaway is that it may be worth ditching the idea of having the user be able to run their code without modification, with all the linking and cross-referencing happening automatically in the background.  Instead, requiring the user to explicitly import and construct an app might lead to a more flexible testing environment, and imply less "hacky" workarounds like a singleton.  Will keep looking into this.

Reply:
> However, the intent is that App gets created in the developer's default code (for example, this file), while the MockLocalHomePlatform gets created in the testing scope. The static instance is the cross-referencing mechanism.

If `smarthome.App` (the stub version) create its own platform, there shouldn't be a need for static cross-reference mecanism (the platform constructor could take the `AppStub` as a parameter and the `AppStub` could ref it as a member after creating it).

In order to trigger intent from test, you could have top-level module function (or `AppStub` methods) that downcast the `smarthome.App` instance in order to get back the platform.

> Yes, but I would need some way of notifying the next App stub of this instance.

Could the next `App` create their own platform?

> I think my takeaway is that it may be worth ditching the idea of having the user be able to run their code without modification, with all the linking and cross-referencing happening automatically in the background. Instead, requiring the user to explicitly import and construct an app might lead to a more flexible testing environment, and imply less "hacky" workarounds like a singleton. Will be mindful of this if it grows into a larger issue.

That's also a good way to think about it, you could always layer the magic back later in this prove too tedious.

Reply:
I pushed a implementation of a top-level module function rather than a singleton.  Closing this thread.